internal/handler/authenticate.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
+	"github.com/FerretDB/FerretDB/internal/handler/common"
+	"github.com/FerretDB/FerretDB/internal/handler/handlererrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/util/password"
+)
+
+// authenticate validates users with stored credentials in admin.systems.user.
+// If EnableNewAuth is false or bypass backend auth is set false, it succeeds
+// authentication and let backend handle it.
+//
+// When admin.systems.user contains no user, the authentication is delegated to
+// the backend. This may change once local exception is implemented.
+func (h *Handler) authenticate(ctx context.Context) error {
+	if !h.EnableNewAuth {
+		return nil
+	}
+
+	if !conninfo.Get(ctx).BypassBackendAuth() {
+		return nil
+	}
+
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	username, pwd := conninfo.Get(ctx).Auth()
+
+	filter := must.NotFail(types.NewDocument("user", username))
+
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	defer qr.Iter.Close()
+
+	var storedUser *types.Document
+
+	var hasUser bool
+
+	for {
+		var v *types.Document
+		_, v, err = qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		hasUser = true
+
+		var matches bool
+
+		if matches, err = common.FilterDocument(v, filter); err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		if matches {
+			storedUser = v
+			break
+		}
+	}
+
+	if !hasUser {
+		// There is no user in the database, let the backend check the authentication.
+		// We do not want unauthenticated users accessing the database, while allowing
+		// users with valid backend credentials to access the database
+		// until local exception is implemented.
+		conninfo.Get(ctx).UnsetBypassBackendAuth()
```

Comment:
Alternatively, we can just let any `PLAIN` user authenticated if there is no user, but I consider that could be dangerous in scenario if all users were dropped. FYI, for `SCRAM` authentication it fails if there is no user.

Reply:
May we add a TODO here for the local exception implementation ticket? It'd be helpful to keep track of it :)

Reply:
I see your point @AlekSi, I'm not particularly pleased with current logic, feel free to suggest if you have a though on a better choice?

Meanwhile I'll add `h.L.DPanic` :+1:

Reply:
With `h.L.DPanic`, the first user creation in integration test setup won't work. I suppose I can create them earlier

Reply:
Let's do h.L.Error instead. Because I want to test this case in integration test, and panic is not ideal for that.

Reply:
Updated #4100